pkg/sanity/cleanup.go

```diff
@@ -36,110 +42,284 @@ type VolumeInfo struct {
 	VolumeID string
 }
 
-// Cleanup keeps track of resources, in particular volumes, which need
-// to be freed when testing is done. All methods can be called concurrently.
+// Cleanup keeps track of resources, in particular volumes and snapshots, that
+// need to be freed when testing is done. It implements both ControllerClient
+// and NodeClient and should be used as the only interaction point to either
+// APIs. That way, Cleanup can ensure that resources are marked for cleanup as
+// necessary.
+// All methods can be called concurrently.
 type Cleanup struct {
-	Context                    *TestContext
-	ControllerClient           csi.ControllerClient
-	NodeClient                 csi.NodeClient
+	Context *TestContext
+	// ControllerClient is meant for struct-internal use only
+	csi.ControllerClient
```

Comment:
It's meant for that, but we don't enforce that because the embedded member is getting exported, right?

I'm on the edge whether the API should prevent access by making it an unexpected member (`controllerClient csi.ControllerClient`). There may be valid cases where a user may want to call the methods that aren't wrapped.

I think I prefer keeping it like this.

Reply:
I left a comment only because unexporting wouldn't help: the (repo-internal) consumers are all part of the same `sanity` package, so they'd still be able to access `controllerClient`. It'd have to be moved into a sub-package, but I didn't want to go that far in my PR.

I can't think of too many reasons why users wouldn't want to go through the `Cleanup` layer: unless csi-test was buggy, it should do the right thing. The one case I can think of is when the `Cleanup()` part shouldn't be executed. Maybe that's reasonable, so let's stick to keeping it exported. I updated the comment to make the implications more clear.